corehq/apps/geospatial/models.py

```diff
@@ -1 +1,130 @@
+
+    def _get_distance_between_points(self, start_point, end_point):
+        """
+        Calculate the great circle distance between two points
+        on the earth (specified in decimal degrees) using the haversine formula
+
+        Returns the distance between two points rounded to two decimal places, in km
+        """
+        # Convert decimal degrees to radians
+        lon1, lat1, lon2, lat2 = map(radians, [start_point.x, start_point.y, end_point.x, end_point.y])
+
+        # Calculate differences between latitudes and longitudes
+        dlon = lon2 - lon1
+        dlat = lat2 - lat1
+
+        # Calculate the square of half the chord length between the points
+        angular_distance = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
+
+        # Calculate the central angle between the points
+        central_angle = 2 * asin(sqrt(angular_distance))
+
+        EARTH_RADIUS_KM = 6371.0
+        distance_km = EARTH_RADIUS_KM * central_angle
+
+        return round(distance_km, 2)
+
+    def get_unassigned_objectives(self, id_only=True):
+        return [
+            (objective.id if id_only else objective) for objective in self.objectives
+            if not objective.is_assigned
+        ]
+
+    def allocate_objectives(self):
+        """
+        Allocates objectives to given users within the constraints of `max_distance` and `max_assignable`.
+        Returns a dict with user_id as the key and a list of assigned objective_id as the value.
+            {
+                'user_a': ['obj_a', 'obj_b']
+            }
+        Users that have not been assigned at least one objective will not be included in the final output.
+        """
+        if not (self.users and self.objectives):
+            return {}
+
+        gdf = self._create_dataframe(self.users)
+
+        user_assignment = {}
+        for obj in self.objectives:
+            # We have run out of available users, and so we cannot assign any more objectives
+            if not len(gdf):
+                break
+
+            obj_point = obj.get_point()
+            users_dist_to_obj = gdf.distance(obj_point)
+
+            closest_user_id = users_dist_to_obj.idxmin()
+            closest_user_geo = gdf.loc[closest_user_id].geometry
+            closest_user_dist = self._get_distance_between_points(closest_user_geo, obj_point)
+
+            # If objective is too far for closest user,
+            # then no user can access this objective and we'll skip it
+            if closest_user_dist > self.max_distance:
+                continue
+
+            if closest_user_id not in user_assignment:
+                user_assignment[closest_user_id] = []
+            user_assignment[closest_user_id].append(obj.id)
+            obj.is_assigned = True
+
+            # Remove user from selection if they've hit their assignable limit
+            if len(user_assignment[closest_user_id]) == self.max_assignable:
```

Comment:
Nit: can we change this to 
```suggestion
            if len(user_assignment[closest_user_id]) >= self.max_assignable:
```
I'm just paranoid about these things...

Reply:
I feel like if we allow for the length of objectives to be more than `max_assignable`, we might potentially run into a situation where users get over-assigned. Is there a particular scenario you're imaginging that this might happen?

Would it maybe be better to have a separate check for `if len(user_assignment[closest_user_id] > self.max_assignable` that we then take appropriate action to reduce it to the max allowed?